helix-term/src/ui/editor.rs

```diff
@@ -735,8 +742,38 @@ impl EditorView {
                 // first execute whatever put us into insert mode
                 self.last_insert.0.execute(cxt);
                 // then replay the inputs
-                for &key in &self.last_insert.1.clone() {
-                    self.insert_mode(cxt, key)
+                for key in self.last_insert.1.clone() {
+                    match key {
+                        InsertEvent::Key(key) => self.insert_mode(cxt, key),
+                        InsertEvent::CompletionApply(compl) => {
+                            let (view, doc) = current!(cxt.editor);
+
+                            doc.restore(view.id);
+
+                            let text = doc.text().slice(..);
+                            let cursor = doc.selection(view.id).primary().cursor(text);
+
+                            let shift_position =
+                                |pos: usize| -> usize { pos + cursor - compl.trigger_offset };
+
+                            let tx = Transaction::change(
+                                doc.text(),
+                                compl
+                                    .transaction
+                                    .changes_iter()
+                                    .filter_map(|(start, end, t)| {
+                                        (start..=end).contains(&compl.trigger_offset).then(|| {
+                                            (shift_position(start), shift_position(end), t)
+                                        })
+                                    }),
+                            );
+                            doc.apply(&tx, view.id);
+                        }
+                        InsertEvent::TriggerCompletion => {
+                            let (_, doc) = current!(cxt.editor);
+                            doc.savepoint();
+                        }
```

Comment:
I don't think this is necessary, it's just for reverting completion changes when you tab through the completion menu. I'd remove `InsertEvent::TriggerCompletion`

Reply:
Actually I think it is. When I remove this, the completion changes are applied after the text typed by the user.
![Peek 2022-02-17 16-02](https://user-images.githubusercontent.com/41059082/154509432-e32d6133-3be1-4197-b39f-e82a8a92a608.gif)

Reply:
Ah right, because you also have that `doc.restore` on L751. You should remove both `savepoint()` and `restore()` then it should work.

Reply:
@archseer it's the same story when I remove both lines. The `save`  and `restore` logic is copied from completion logic. Otherwise it doesn't work.

Reply:
Ah right. This is what happens:

`std::` <- completion triggers, doc.savepoint() is called
`std::thre` <- filters completion items
`Enter` <- clears back to save point `std::` then applies the completion.

I'll do some testing to see if I can avoid that.